### ISSUE TITLE
JSON deserialization fails in iOS release build if class implements an interface

### ISSUE THREAD

**Issue by markuspalme** [dotnet/runtime#75802]
### Description

A simple JSON deserialization using `Newtonsoft.Json` fails on iOS in a release build as soon as the class that is trying be be de-serialized implements an interface. 

The same app works in Debug build in the iOS simulator as expected.

### Reproduction Steps

1. Create a new iOS app with `dotnet new ios`
2. Set a valid bundle identifier that allows installing the app on a physical device
3. Add this code to the project:
  
    ```
    public class ProductImage : IEntity
    {
       public string ProductNumber { get; set; }
       public Guid Id { get; set; }
    }
      
    public interface IEntity
    {
       Guid Id { get; set; }
    }
    ```

    During startup, run this code:
    
      ```
      var x = @"{
         ""productNumber"": ""P1"",
         ""id"": ""46c67d7c-fd15-4d89-9fed-991c48c1bacf""
      }";
      
      var pi = JsonConvert.DeserializeObject<ProductImage>(x);
      
      ```

4. Build the app for a device:  `dotnet publish testapp.csproj -f:net6.0-ios -c:Release /p:CodesignKey="Apple Development: ******" /p:CodesignProvision="Test app" /p:ArchiveOnBuild=true -r ios-arm64 --self-contained`

5. Deploy to device and run

Full project file:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0-ios</TargetFramework>
    <OutputType>Exe</OutputType>
    <Nullable>enable</Nullable>
    <ImplicitUsings>true</ImplicitUsings>
    <SupportedOSPlatformVersion>13.0</SupportedOSPlatformVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
  </ItemGroup>
</Project>
```

Self-contained example project:
https://github.com/markuspalme/dotnetruntime-75802

### Expected behavior

The app starts successfully and manages to de-serialize the JSON regardless of whether the class implements an interface or not. 

Note that removing the interface will make the de-serialization work as expected.

### Actual behavior

JSON deserialization fails with this exception:

`Error setting value to 'Id' on 'testapp.ProductImage'.`

### Regression?

The same code works in Xamarin.iOS.

### Configuration

.NET 6.0.401
iOS 16, same on 15.6.1

**Comment by gregsdennis** [comment]
Are you certain this is a bug in .Net?  Seems like it's a Newtonsoft issue.  If so, the bug should be filed in that repo.

**Comment by markuspalme** [comment]
@gregsdennis It works just fine in `net6` on Windows, MacOS and also in the iOS simulator - all with the same `Newtonsoft.Json` package. So I think this is a runtime or maybe AOT issue (iOS requires AOT).

**Comment by gregsdennis** [comment]
Understood, but it might be that the package just needs handle a special case in this OS.  There is the possibility that the issue is the package.

**Comment by markuspalme** [comment]
Unlikely, the same code works with Xamarin.iOS.

**Comment by Cheesebaron** [comment]
@markuspalme could it be the linker stripping out stuff?

Could you try add `<PublishTrimmed>false</PublishTrimmed>` in your Project Properties and see if that helps?

**Comment by markuspalme** [comment]
I have tried `System.Text.Json` instead of `Newtonsoft.Json`. It does not throw an exception, but it does not set any property values.

```
var pi = System.Text.Json.JsonSerializer.Deserialize<ProductImage>(x);
```

**Comment by markuspalme** [comment]
Here's a small self-contained example:
https://github.com/markuspalme/dotnetruntime-75802

**Comment by markuspalme** [comment]
Adding ` <UseInterpreter>true</UseInterpreter>` as suggested in this issue helps - but I don't think this should be neccesary:
https://github.com/xamarin/xamarin-macios/issues/15961

**Comment by Cheesebaron** [comment]
@markuspalme UseInterpreter is not for Release builds, Apple won't allow that.

**Comment by markuspalme** [comment]
@Cheesebaron Yes, I figured but wanted to share that it helps here - maybe it helps pinpointing the root cause.

**Comment by jkotas** [comment]
System.Text.Json should be the right area label. 

Source generators are the only 100% reliable way to perform Json serialization/deserialization in the presence of trimming or AOT compilation without fallbacks.

Unfortunately, trim warnings that would notify about this problem are disabled for iOS apps by default. You can set `<SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>`  in your .csproj to enable trim warnings to see all potential places in your app that can break due to trimming.

This is duplicate https://github.com/dotnet/runtime/issues/74141 and number of other similar issues.

@eiriktsarpalis My suggestion would be:
- Make sure that we have documentation that we can point to and resolve these issues against. Do we have a documentation that says "System.Text.Json source generators are the only 100% reliable way to perform Json serialization/deserialization in the presence of trimming"? 
- Note the feedback on issues that track enabling the link warning in app-model specific SDKs. It is https://github.com/xamarin/xamarin-macios/issues/11246 for iOS.

**Comment by eiriktsarpalis** [comment]
> System.Text.Json should be the right area label.

The OP concerns Json.NET (Newtonsoft) serialization issues in iOS. I agree that it falls under the same category as #74141, but I wonder what our approach should be when users report trimming-related failures when using third party reflection-based libraries. We might want to consider asking them to file an issue with said third-party libraries asking about AOT support.

> Do we have a documentation that says "System.Text.Json source generators are the only 100% reliable way to perform Json serialization/deserialization in the presence of trimming"?

[It is](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-source-generation-modes?pivots=dotnet-6-0), although the wording is certainly more understated than that.

**Comment by markuspalme** [comment]
@jkotas I will try to enable the trim warnings.

JSON is at the heart of many apps and this feels like a regression coming from Xamarin.iOS where such scenarios worked out of the box. Migrating to`System.Text.Json` source generators is a possibility for user code, but many libraries depend on it.

**Comment by markuspalme** [comment]
@eiriktsarpalis Thanks for your input. Given that earlier versions of Xamarin.iOS supported this scenario out of the box even in AOT mode, this seems like a bad surprise or even a blocker for anyone porting applications to `net6-ios`.

**Comment by markuspalme** [comment]
@eiriktsarpalis So my understanding is that this problem here is caused by trimming. To preserve all the types in my code, I have added this to the project file as described here (https://devblogs.microsoft.com/dotnet/customizing-trimming-in-net-core-5/):

```
<ItemGroup>
  <TrimmerRootDescriptor Include="TrimmerRoots.xml" />
</ItemGroup>
```

The `TrimmerRoots.xml` looks like this:

```
<?xml version="1.0" encoding="utf-8"?>

<linker>
    <assembly fullname="testapp" preserve="all" />
</linker>
```

The de-serialization does not work with that in place. Shouldn't this ensure that all my types are preserved and available through reflection?

The whole story around trimming and linking is rather intimidating:

- there is the `[Preserve]` attribute 
- there is the before mentioned external XML configuration option
- For `net6-ios` projects there is the `MtouchLink` property in the project file 

Is there any guidance available on which one to use when and how they work together?

**Comment by vitek-karas** [comment]
The trimming options are described here: https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-6-0

The `Preserve` attribute has been deprecated... I don't actually think it works anymore (but maybe it does for iOS targets)

The trimmer should produce warnings if the XML is not understood - but mobile targets disable these warnings by default (there are just way too many warnings produced by all layers of the system currently). You can turn it on by setting `SuppressTrimAnalysisWarnings=false` and look for warnings pointing to your XML file.

@sbomer for additional details.

**Comment by markuspalme** [comment]
@vitek-karas I did that and get the expected warning about `Newtonsoft.Json`. But all types that are involded in the de-serialization are marked for preservation in the XML file.

There are no warnings regarding the XML file. You can access the full project here:
https://github.com/markuspalme/dotnetruntime-75802

**Comment by vitek-karas** [comment]
@sbomer - could you please try to take a look if the XML descriptor works as expected in this case?

**Comment by markuspalme** [comment]
@sbomer here you go:
[build.binlog.zip](https://github.com/dotnet/runtime/files/9610291/build.binlog.zip)

**Comment by markuspalme** [comment]
@sbomer Did the binlog reveal anything interesting?

**Comment by steveisok** [comment]
@sbomer if you need to bounce ideas off of someone who has ios set up, please let me know.

**Comment by markuspalme** [comment]
@sbomer Here you go: https://github.com/markuspalme/dotnetruntime-75802/blob/master/repro.zip

**Comment by markuspalme** [comment]
@sbomer Here are the IPA files:

With descriptor:
[testapp-descriptor.ipa.zip](https://github.com/dotnet/runtime/files/9713950/testapp-descriptor.ipa.zip)

Without descriptor:
[testapp-without-descriptor.ipa.zip](https://github.com/dotnet/runtime/files/9713953/testapp-without-descriptor.ipa.zip)

There is is at least a difference in size:

<img width="740" alt="image" src="https://user-images.githubusercontent.com/1481801/194014326-371d9036-5f5c-438a-8c19-15c921f32a44.png">

**Comment by markuspalme** [comment]
@sbomer I saw the same but was not sure it's relevant because I don't know if the assembly is used for anything but metadata in the AOT build. 

The error is not reproducible on the simulator. I do think it's an AOT issue and not a trimming issue for this reason - but I have no idea how to investigate any further by myself.

**Comment by MichalStrehovsky** [comment]
Cc @akoeplinger for ILStrip

**Comment by markuspalme** [comment]
@MichalStrehovsky I did that. The assembly is now intact:

<img width="805" alt="image" src="https://user-images.githubusercontent.com/1481801/194268044-cba11f53-8eda-4dc0-abd1-e7a842637690.png">

But the error still occurs on device. As far as I know, the IL is used for metadata only in the AOT only which is probably why method bodies are stripped away.

**Comment by markuspalme** [comment]
@MichalStrehovsky  Is there anything else I can help with analyzing this issue?

**Comment by markuspalme** [comment]
@MichalStrehovsky @steveisok @sbomer Could this be related to #69410?

**Comment by markuspalme** [comment]
Can still be reproduced with .NET 7.

**Comment by minglu** [comment]
following

**Comment by steveisok** [comment]
@akoeplinger please investigate if this is indeed an ILStrip issue.

**Comment by steveisok** [comment]
@akoeplinger do you the recent ILStrip changes might have fixed this? Can you please validate?

**Comment by markuspalme** [comment]
This is fixed in 8.0.100-preview.7.23376.3, presumably by #69410.
